rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+header value on a `300` series redirect.
+
+The Queue will now store only the name of Artifacts and would treat the name of
+an Artifact as an opaque identifier and defer all management of the actual
+bytes to this system.  The Queue would also stop doing IP to Object store
+mappings.
+
+This RFC is to decide on what the Artifacts service is and its interface, not
+implementation details
+
+## Motivation
+
+* We want the Queue to be easier to understand, develop and maintain
+* We want Artifacts to be easier to understand, develop and maintain
+* We want better Artifact handling without increasing complexity of the Queue
+  codebase
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of Artifacts into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
+# Details
+
+* The Artifact service will start from the Queue's `artifacts.js` file
+* Auth might not be done with scopes and tc-auth, but a shared secret between
+  Queue and Artifacts
+* The Queue will block new tasks after task resolution by not issuing new
+  signed urls
+* Data storage in Artifacts will be done with Postgres and not Azure Table
+  Storage
+* We do not intend to have an operation to list artifacts in the API
+* The redirecting logic of the Artifact service will adhere to the HTTP spec
+  using 300-series redirects where possible
+* Where content-negotiation is impossible, the service will block creation of
+  artifacts when incoming requests aren't compatible with required response
+* We will support uploading artifacts to different regions so that we can
+
+## API
+This API is up for debate and is only an illustrative starting point.  
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  This will be an object shaped like
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key -> list of strings)
+* All query string, path, fragements are contained in a fully formed `url` property
+
+### Creation
+
+```
+PUT /artifacts/:name
+```
+
+This endpoint will take an origin and name.  It will return a `200` with a
+document which contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the artifact in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
```

Comment:
That's an implementation detail we should finalize as we progress.  I want to spend a bit of time with other object stores before we solidify multipart uploads across services.

Reply:
I think the public API is one of the things to settle before we start implementing this.

Reply:
If this will match the existing blob storage type, it might be OK to just refer to that -- if that's sufficiently general to support other clouds' multipart implementations.